File: cognipy/CogniPyLib/JenaRules/ReactiveRuleCompiler.cs
using Microsoft.CSharp;
using Ontorion.CNL.DL;
using System;
using System.CodeDom.Compiler;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Text;

namespace OWLServices
{

    public class ReactiveRuleCompiler
    {
        private const string classTemplate = @"
            using System;
            using System.Linq;
            using System.Collections.Generic;
            using System.Linq.Expressions;
            
            public class Rule
            {{
                public void Execute(object[] Matches, Action<HashSet<string>>  knowledgeInsert,  Action<HashSet<string>> knowledgeDelete, Action<int, HashSet<string>> delayedKnowledgeInsert) {{
                    /*{1}*/
                    this.DlDelayedKnowledgeInsert = delayedKnowledgeInsert;
                    this.DlKnowledgeDelete= knowledgeDelete;
                    this.DlKnowledgeInsert= knowledgeInsert;
                    {2};
                    {0};
                }}

                public dynamic Ontorion;
                public dynamic Factory;
                public dynamic Outer;
                public dynamic Executor;
                public Action<int, HashSet<string>> DlDelayedKnowledgeInsert;
                public Action<HashSet<string>> DlKnowledgeInsert;
                public Action<HashSet<string>> DlKnowledgeDelete;

                public string ID(string dl) {{ return Ontorion.ID(dl); }}
                public string EN(string en){{return Ontorion.EN(en);}}
                public string InstanceDL(string en){{return Ontorion.InstanceDL(en);}}
                public void SetProgress(double completed){{Ontorion.SetProgress(completed);}}
                public void WriteMessage(int priority, string message){{Ontorion.WriteMessage(priority,message);}}
                public void WriteMessage(string message){{Ontorion.WriteMessage(0,message);}}
                public void WriteDebugMessage(string message){{Ontorion.WriteDebugMessage(message);}}
                public string UniqueID(){{return Ontorion.UniqueID();}}

                public static void BreakPoint(){{
                    System.Diagnostics.Debugger.Launch();
                    System.Diagnostics.Debugger.Break();
                }}

                public void Move(string agentId, string agentStateId, string messageId,
                    string messageTitle, Func<string> task)
                {{
                    Ontorion.AgentSupport.Move(agentId,agentStateId,messageId,messageTitle,task);
                }}

                public void Once(string uniqueness, Action task)
                {{
                    Ontorion.AgentSupport.Once(uniqueness,task);
                }}

                public dynamic External(string name)
                {{
                    return Factory.External(name,this);
                }}

                public void KnowledgeInsertWithDelay(string knowledge, int timeout_ms)
                {{
                    DlDelayedKnowledgeInsert(timeout_ms, Ontorion.KnowledgeSplit(knowledge));
                }}

                public void KnowledgeDelete(string knowledge)
                {{
                    DlKnowledgeDelete(Ontorion.KnowledgeSplit(knowledge));
                }}

                public void KnowledgeInsert(string knowledge)
                {{
                    DlKnowledgeInsert(Ontorion.KnowledgeSplit(knowledge));
                }}

                public string CreateMessage(string origin, string header, string body)
                {{
                    return Ontorion.AgentSupport.CreateMessage(origin, header,body);
                }}

                public void CreateAgent(string agentid, string state)
                {{
                    Ontorion.AgentSupport.CreateAgent(agentid,state);
                }}

                public string GetMessageBody(string messageId)
                {{
                    return Ontorion.AgentSupport.GetMessageBody(messageId);
                }}

                public string GetMessageHeader(string messageId)
                {{
                    return Ontorion.AgentSupport.GetMessageHeader(messageId);
                }}
            }}
        ";

        private static Dictionary<string, Type> TypeCache = new Dictionary<string, Type>();

        public static Type LoadRuleType(string rule, string ruleHead, List<IExeVar> vars)
        {
            lock (TypeCache)
            {
                if (!TypeCache.ContainsKey(rule))
                {
                    StringBuilder sb = new StringBuilder();
                    Dictionary<string, int> maxnos = new Dictionary<string, int>();
                    foreach (var v in vars)
                    {
                        if (v.isVar())
                        {
                            var dpos = (v as ISwrlVar).getVar().LastIndexOf('-');
                            var k = (v as ISwrlVar).getVar().Substring(0, dpos);
                            var l = (v as ISwrlVar).getVar().Substring(dpos + 1);
                            if (l[0] == 'x' || l[0] == '0')
                            {
                                if (l[0] == '0')
                                {
                                    var idx = int.Parse(l);
                                    if (!maxnos.ContainsKey("?" + k))
                                        maxnos.Add("?" + k, idx);
                                    else
                                        maxnos["?" + k] = Math.Max(maxnos["?" + k], idx);
                                }
                                else
                                {
                                    if (!maxnos.ContainsKey("?" + k))
                                        maxnos.Add("?" + k, 1);
                                }
                            }
                            else
                            {
                                var idx = int.Parse(l);
                                if (!maxnos.ContainsKey(k))
                                    maxnos.Add(k, idx);
                                else
                                    maxnos[k] = Math.Max(maxnos[k], idx);
                            }
                        }
                    }
                    sb.AppendLine("var __menum=Matches.GetEnumerator();");
                    HashSet<string> nosd = new HashSet<string>();
                    foreach (var v in vars)
                    {
                        if (v.isVar())
                        {
                            sb.AppendLine("__menum.MoveNext();");
                            var dpos = (v as ISwrlVar).getVar().LastIndexOf('-');
                            var k = (v as ISwrlVar).getVar().Substring(0, dpos);
                            var l = (v as ISwrlVar).getVar().Substring(dpos + 1);
                            if (l[0] == 'x' || l[0] == '0')
                            {
                                if (!nosd.Contains("?" + k))
                                {
                                    if ((l[0] == 'x' && maxnos["?" + k] == 1) || (l[0] == '0' && maxnos["?" + k] == int.Parse(l)))
                                    {
                                        sb.Append("var ");
                                        sb.Append(k.Replace("-", "_"));
                                        if (v is SwrlIVar)
                                            sb.AppendLine(" = ID(__menum.Current as string) ;");
                                        else
                                            sb.AppendLine(" = __menum.Current;");
                                        nosd.Add("?" + k);
                                    }
                                }
                            }
                            else
                            {
                                if (!nosd.Contains(k))
                                {
                                    sb.Append("var ");
                                    sb.Append(k.Replace("-", "_"));
                                    if (v is SwrlIVar)
                                        sb.AppendLine("= new Dictionary<int,string>();");
                                    else
                                        sb.AppendLine("= new Dictionary<int,dynamic>();");
                                    nosd.Add(k);
                                }
                                var idx = int.Parse(l);
                                if (v is SwrlIVar)
                                    sb.Append(k + ".Add(" + idx.ToString() + ", ID(__menum.Current as string));");
                                else
                                    sb.Append(k + ".Add(" + idx.ToString() + ", __menum.Current);");
                            }
                        }
                    }
                    var classSource = string.Format(classTemplate, rule, ruleHead, sb.ToString());
                    var assembly = CompileAssembly(classSource);
                    var type = assembly.GetType("Rule");
                    TypeCache[rule] = type;
                }
                return TypeCache[rule];
            }
        }

        private static Assembly CompileAssembly(string source)
        {
            var compilerParameters = new CompilerParameters()
            {
                GenerateExecutable = false,
                GenerateInMemory = true,
                ReferencedAssemblies =
                {
                    "System.dll", // needed for linq + expressions to compile
                    "System.Core.dll", // needed for linq + expressions to compile
                    "System.Data.dll", // needed for linq + expressions to compile
                    "Microsoft.CSharp.dll", // needed for linq + expressions to compile
                },
                IncludeDebugInformation = true,
            };

            string[] References = {
                };

            foreach (var reference in References)
            {
                compilerParameters.ReferencedAssemblies.Add(AppDomain.CurrentDomain.BaseDirectory
                           + string.Format("\\{0}.dll", reference));
            }

            var compileProvider = new CSharpCodeProvider();

            var fileName = Path.GetFullPath(Path.Combine(Path.GetTempPath(), Path.GetTempFileName() + ".cs"));
            File.WriteAllText(fileName, source);
            var results = compileProvider.CompileAssemblyFromFile(compilerParameters, fileName);

            //            var results = compileProvider.CompileAssemblyFromSource(compilerParameters, source);
            if (results.Errors.HasErrors)
            {
                StringBuilder msg = new StringBuilder();
                int idx = 1;
                foreach (CompilerError err in results.Errors)
                {
                    msg.AppendFormat("{1}[{2}] : {0}", err.ErrorText, err.IsWarning ? "WARNING" : "ERROR", idx++);
                    msg.AppendLine();
                }
                throw new RuleCompilationException(msg.ToString());
            }

            var assembly = results.CompiledAssembly;
            return assembly;
        }
    }
    public class RuleCompilationException : Exception
    {
        public RuleCompilationException(string msg) : base(msg)
        {
        }
    }
}

File: cognipy/cognipylib/owlservices/ReactiveRuleCompiler.java
package owlservices;

import cognipy.cnl.dl.*;
import java.util.*;
import java.io.*;
import java.nio.file.*;

public class ReactiveRuleCompiler
{
	private static final String classTemplate = "" + "\r\n" +
"            using System;" + "\r\n" +
"            using System.Linq;" + "\r\n" +
"            using System.Collections.Generic;" + "\r\n" +
"            using System.Linq.Expressions;" + "\r\n" +
"            " + "\r\n" +
"            public class Rule" + "\r\n" +
"            {{" + "\r\n" +
"                public void Execute(object[] Matches, Action<HashSet<string>>  knowledgeInsert,  Action<HashSet<string>> knowledgeDelete, Action<int, HashSet<string>> delayedKnowledgeInsert) {{" + "\r\n" +
" /*{1}*/" + "\r\n" +
"                    this.DlDelayedKnowledgeInsert = delayedKnowledgeInsert;" + "\r\n" +
"                    this.DlKnowledgeDelete= knowledgeDelete;" + "\r\n" +
"                    this.DlKnowledgeInsert= knowledgeInsert;" + "\r\n" +
"                    {2};" + "\r\n" +
"                    {0};" + "\r\n" +
"                }}" + "\r\n" +
"\r\n" +
"                public dynamic Ontorion;" + "\r\n" +
"                public dynamic Factory;" + "\r\n" +
"                public dynamic Outer;" + "\r\n" +
"                public dynamic Executor;" + "\r\n" +
"                public Action<int, HashSet<string>> DlDelayedKnowledgeInsert;" + "\r\n" +
"                public Action<HashSet<string>> DlKnowledgeInsert;" + "\r\n" +
"                public Action<HashSet<string>> DlKnowledgeDelete;" + "\r\n" +
"\r\n" +
"                public string ID(string dl) {{ return Ontorion.ID(dl); }}" + "\r\n" +
"                public string EN(string en){{return Ontorion.EN(en);}}" + "\r\n" +
"                public string InstanceDL(string en){{return Ontorion.InstanceDL(en);}}" + "\r\n" +
"                public void SetProgress(double completed){{Ontorion.SetProgress(completed);}}" + "\r\n" +
"                public void WriteMessage(int priority, string message){{Ontorion.WriteMessage(priority,message);}}" + "\r\n" +
"                public void WriteMessage(string message){{Ontorion.WriteMessage(0,message);}}" + "\r\n" +
"                public void WriteDebugMessage(string message){{Ontorion.WriteDebugMessage(message);}}" + "\r\n" +
"                public string UniqueID(){{return Ontorion.UniqueID();}}" + "\r\n" +
"\r\n" +
"                public static void BreakPoint(){{" + "\r\n" +
"                    System.Diagnostics.Debugger.Launch();" + "\r\n" +
"                    System.Diagnostics.Debugger.Break();" + "\r\n" +
"                }}" + "\r\n" +
"\r\n" +
"                public void Move(string agentId, string agentStateId, string messageId," + "\r\n" +
"                    string messageTitle, Func<string> task)" + "\r\n" +
"                {{" + "\r\n" +
"                    Ontorion.AgentSupport.Move(agentId,agentStateId,messageId,messageTitle,task);" + "\r\n" +
"                }}" + "\r\n" +
"\r\n" +
"                public void Once(string uniqueness, Action task)" + "\r\n" +
"                {{" + "\r\n" +
"                    Ontorion.AgentSupport.Once(uniqueness,task);" + "\r\n" +
"                }}" + "\r\n" +
"\r\n" +
"                public dynamic External(string name)" + "\r\n" +
"                {{" + "\r\n" +
"                    return Factory.External(name,this);" + "\r\n" +
"                }}" + "\r\n" +
"\r\n" +
"                public void KnowledgeInsertWithDelay(string knowledge, int timeout_ms)" + "\r\n" +
"                {{" + "\r\n" +
"                    DlDelayedKnowledgeInsert(timeout_ms, Ontorion.KnowledgeSplit(knowledge));" + "\r\n" +
"                }}" + "\r\n" +
"\r\n" +
"                public void KnowledgeDelete(string knowledge)" + "\r\n" +
"                {{" + "\r\n" +
"                    DlKnowledgeDelete(Ontorion.KnowledgeSplit(knowledge));" + "\r\n" +
"                }}" + "\r\n" +
"\r\n" +
"                public void KnowledgeInsert(string knowledge)" + "\r\n" +
"                {{" + "\r\n" +
"                    DlKnowledgeInsert(Ontorion.KnowledgeSplit(knowledge));" + "\r\n" +
"                }}" + "\r\n" +
"\r\n" +
"                public string CreateMessage(string origin, string header, string body)" + "\r\n" +
"                {{" + "\r\n" +
"                    return Ontorion.AgentSupport.CreateMessage(origin, header,body);" + "\r\n" +
"                }}" + "\r\n" +
"\r\n" +
"                public void CreateAgent(string agentid, string state)" + "\r\n" +
"                {{" + "\r\n" +
"                    Ontorion.AgentSupport.CreateAgent(agentid,state);" + "\r\n" +
"                }}" + "\r\n" +
"\r\n" +
"                public string GetMessageBody(string messageId)" + "\r\n" +
"                {{" + "\r\n" +
"                    return Ontorion.AgentSupport.GetMessageBody(messageId);" + "\r\n" +
"                }}" + "\r\n" +
"\r\n" +
"                public string GetMessageHeader(string messageId)" + "\r\n" +
"                {{" + "\r\n" +
"                    return Ontorion.AgentSupport.GetMessageHeader(messageId);" + "\r\n" +
"                }}" + "\r\n" +
"            }}" + "\r\n" +
"        ";

	private static HashMap<String, java.lang.Class> TypeCache = new HashMap<String, java.lang.Class>();

	public static java.lang.Class LoadRuleType(String rule, String ruleHead, ArrayList<IExeVar> vars)
	{
		synchronized (TypeCache)
		{
			if (!TypeCache.containsKey(rule))
			{
				StringBuilder sb = new StringBuilder();
				HashMap<String, Integer> maxnos = new HashMap<String, Integer>();
				for (IExeVar v : vars)
				{
					if (v.isVar())
					{
						int dpos = (v instanceof ISwrlVar ? (ISwrlVar)v : null).getVar().lastIndexOf('-');
						String k = (v instanceof ISwrlVar ? (ISwrlVar)v : null).getVar().substring(0, dpos);
						String l = (v instanceof ISwrlVar ? (ISwrlVar)v : null).getVar().substring(dpos + 1);
						if (l.charAt(0) == 'x' || l.charAt(0) == '0')
						{
							if (l.charAt(0) == '0')
							{
								int idx = Integer.parseInt(l);
								if (!maxnos.containsKey("?" + k))
								{
									maxnos.put("?" + k, idx);
								}
								else
								{
									maxnos.put("?" + k, Math.max(maxnos.get("?" + k), idx));
								}
							}
							else
							{
								if (!maxnos.containsKey("?" + k))
								{
									maxnos.put("?" + k, 1);
								}
							}
						}
						else
						{
							int idx = Integer.parseInt(l);
							if (!maxnos.containsKey(k))
							{
								maxnos.put(k, idx);
							}
							else
							{
								maxnos.put(k, Math.max(maxnos.get(k), idx));
							}
						}
					}
				}
				sb.append("var __menum=Matches.GetEnumerator();" + "\r\n");
				HashSet<String> nosd = new HashSet<String>();
				for (IExeVar v : vars)
				{
					if (v.isVar())
					{
						sb.append("__menum.MoveNext();" + "\r\n");
						int dpos = (v instanceof ISwrlVar ? (ISwrlVar)v : null).getVar().lastIndexOf('-');
						String k = (v instanceof ISwrlVar ? (ISwrlVar)v : null).getVar().substring(0, dpos);
						String l = (v instanceof ISwrlVar ? (ISwrlVar)v : null).getVar().substring(dpos + 1);
						if (l.charAt(0) == 'x' || l.charAt(0) == '0')
						{
							if (!nosd.contains("?" + k))
							{
								if ((l.charAt(0) == 'x' && maxnos.get("?" + k).equals(1)) || (l.charAt(0) == '0' && maxnos.get("?" + k).equals(Integer.parseInt(l))))
								{
									sb.append("var ");
									sb.append(k.replace("-", "_"));
									if (v instanceof SwrlIVar)
									{
										sb.append(" = ID(__menum.Current as string) ;" + "\r\n");
									}
									else
									{
										sb.append(" = __menum.Current;" + "\r\n");
									}
									nosd.add("?" + k);
								}
							}
						}
						else
						{
							if (!nosd.contains(k))
							{
								sb.append("var ");
								sb.append(k.replace("-", "_"));
								if (v instanceof SwrlIVar)
								{
									sb.append("= new Dictionary<int,string>();" + "\r\n");
								}
								else
								{
									sb.append("= new Dictionary<int,dynamic>();" + "\r\n");
								}
								nosd.add(k);
							}
							int idx = Integer.parseInt(l);
							if (v instanceof SwrlIVar)
							{
								sb.append(k + ".Add(" + String.valueOf(idx) + ", ID(__menum.Current as string));");
							}
							else
							{
								sb.append(k + ".Add(" + String.valueOf(idx) + ", __menum.Current);");
							}
						}
					}
				}
				String classSource = String.format(classTemplate, rule, ruleHead, sb.toString());
				Assembly assembly = CompileAssembly(classSource);
//C# TO JAVA CONVERTER TODO TASK: There is no equivalent to implicit typing in Java unless the Java 10 inferred typing option is selected:
				var type = assembly.GetType("Rule");
				TypeCache.put(rule, type);
			}
			return TypeCache.get(rule);
		}
	}

	private static Assembly CompileAssembly(String source)
	{
		CompilerParameters compilerParameters = new CompilerParameters();
		compilerParameters.setGenerateExecutable(false);
		compilerParameters.setGenerateInMemory(true);
		compilerParameters.setReferencedAssemblies({"System.dll", "System.Core.dll", "System.Data.dll", "Microsoft.CSharp.dll"});
		compilerParameters.setIncludeDebugInformation(true);

		String[] References = { };

		for (String reference : References)
		{
			compilerParameters.ReferencedAssemblies.Add(AppDomain.CurrentDomain.BaseDirectory + String.format("\\%1$s.dll", reference));
		}

		CSharpCodeProvider compileProvider = new CSharpCodeProvider();

		String fileName = (new File(Paths.get(Path.GetTempPath()).resolve(Path.GetTempFileName() + ".cs").toString())).getAbsolutePath();
		Files.writeString(fileName, source);
		System.CodeDom.Compiler.CompilerResults results = compileProvider.CompileAssemblyFromFile(compilerParameters, fileName);

		//            var results = compileProvider.CompileAssemblyFromSource(compilerParameters, source);
		if (results.Errors.HasErrors)
		{
			StringBuilder msg = new StringBuilder();
			int idx = 1;
			for (CompilerError err : results.Errors)
			{
				msg.append(String.format("%2$s[%3$s] : %1$s", err.ErrorText, err.IsWarning ? "WARNING" : "ERROR", idx++));
				msg.append("\r\n");
			}
			throw new RuleCompilationException(msg.toString());
		}

		System.Reflection.Assembly assembly = results.CompiledAssembly;
		return assembly;
	}
}